### ISSUE TITLE
Road to v0.4

### ISSUE THREAD

**Issue by rtavenar** [tslearn-team/tslearn#221]
Hi there,

We already have a bunch of new things ready (or close to). So I'd suggest we release a v0.4 in the near future (like in 2-3 weeks or so).

Here is a list of the new things to come:
* [x] Early classification method developed in #198 (should be ready in 2-3 weeks I think)
* [x] DTW with custom ground metric developed in #219 (it's already in a good shape, a few minor changes are required)
* [x] Better checks for test data in #223  and bugfix for related #220 issue (looks like it could be done in a short time, but I might be wrong)
* [x] I also think #211 should be merged to deal with both recent and old versions of the sklearn API
* [x] **EDIT**: #222 that adds a new kmeans++ that is supposed to draw better seeds for kmeans when DTW/softDTW is used

There are also the things that are [already merged in `dev`](https://github.com/rtavenar/tslearn/blob/dev/CHANGELOG.md#towards-v040).

Also, I changed some stuff in the docs so that all categories are directly accessible in the top menu bar (and the examples in the gallery are organized into categories), let me know if you have comments on this.

Finally, this v0.4 could be the occasion to change a few things in the organization of the project. Here is a list of what I have in mind:
* [x] Drop Python2 support (should it be dropped for this version already or just warned that it will be dropped in version XX to come?) -> this refers to #191 
* [x] Have better branch names and a default branch that makes sense (I think we should have branch names for past versions -- already pushed to PyPI -- and keep `master` as the default branch that stores changes that are not yet deployed on PyPI) -> this refers to a discussion by @rth in #190  )
* [x] Maybe we should move the docs to `docs/` rather than `tslearn/docs` so that we do not have to exclude this path from the tests -> idea by @rth once again

Finally, you may have noticed that we now have [automatic wheel generation using GitHub Actions](https://github.com/rtavenar/tslearn/actions?query=workflow%3A%22Build+dist+and+wheels%22). This is not really related to v0.4 since I already tested it on the current version, but we should now be able to distribute wheels for a wide range of platforms.

OK, so the idea of this Issue is to discuss these points that we should include in the next release, all your ideas are welcome as soon as they can be implemented in a short amount of time :)

Cc @GillesVandewiele @johannfaouzi

**Comment by rtavenar** [comment]
And maybe it would be time to move from rtavenar / tslearn to tslearn / tslearn

**Comment by johannfaouzi** [comment]
I agree with all of the following suggestions:
> * Drop Python2 support (should it be dropped for this version already or just warned that it will be dropped in version XX to come?) -> this refers to #191

> *  Have better branch names and a default branch that makes sense (I think we should have branch names for past versions -- already pushed to PyPI -- and keep master as the default branch that stores changes that are not yet deployed on PyPI) -> this refers to a discussion by @rth in #190 )

>  * Maybe we should move the docs to docs/ rather than tslearn/docs so that we do not have to exclude this path from the tests -> idea by @rth once again

Also, on this one:
> And maybe it would be time to move from rtavenar / tslearn to tslearn / tslearn

I don't have a strong opinion, I think it could stay here. But I think that it may be ill-timed with the JMLR submission (we can still make changes to the submission though).

**Comment by rtavenar** [comment]
Yep, for the last point, the idea was: if we want to do it, we should do it before providing a camera-ready version, so maybe doing it soon would be a good idea.

Also, about Python 2 support, so you have an opinion, does it feel ok to drop it starting from tslearn v0.4 (and hence stop testing it / delivering wheels & conda packages for it)?

**Comment by GillesVandewiele** [comment]
It could perhaps be a bit too drastic to remove the support straight away. Perhaps we could try to detect whether python2 or python3 is used and raise a deprecation warning for python2 (stating that it will be removed in a later version).

Side note: I didn't manage to review your PR yesterday Romain, I hope I can do it today!

**Comment by rth** [comment]
> It could perhaps be a bit too drastic to remove the support straight away. Perhaps we could try to detect whether python2 or python3 is used and raise a deprecation warning for python2 (stating that it will be removed in a later version).

PyPi stats are not too reliable (bots etc), but they currently show that the vast majority of usage is in Python 3. https://pypistats.org/packages/tslearn

![Screenshot_2020-04-29 PyPI Download Stats](https://user-images.githubusercontent.com/630936/80578627-0f0bed00-8a09-11ea-9b53-f4cffcba332f.png)

So I don't think it would be too unexpected to drop it in 0.4 considering that most of other main packages in the python ecosystem already did it. Or in 0.5 if you prefer to wait one more release. Note that CI is not running python 2 now I think.

**Comment by GillesVandewiele** [comment]
Ok then we could definitely drop it! And perhaps we could still raise a warning for python2 users (that we do not support it and that they better use python3).

**Comment by rth** [comment]
>  we could still raise a warning for python2 users (that we do not support it and that they better use python3).

Recent versions of pip already raises a deprecation warning when one installs anything on Python 2. E.g. with pip 19.3.1,
```
$ pip install numpy
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
[...]
```

I think the correct way to do this is to specify `python_requires=">=3.5"` in setup.py. Then Python 2 users can still install tslearn without error (it will just not be the last version, since pip will not select it). We also have to remember to update the conda recipe with an equivalent apporach. If we don't do this, running `pip install tslearn` would break for Python 2 users after a Python 3 only release.

**Comment by rtavenar** [comment]
OK, so I think we can drop python 2 compatibility for tslearn v0.4. I will follow your guidelines @rth , thanks a lot for that.

And the plan is then to do these things sequentially (what follows is mainly a note for myself):

1. merge #211 and #223 to `dev`
2. create a branch `v0.3` from `master` and setup RTD to generate docs for this branch rather than `master`
3. merge `dev` to `master` that would then become the development branch
4. merge  #198 and #219 to `master` + a PR that would drop Python 2 support when possible
5. Move the `docs/` folder
6. Release v0.4 (ie create a branch `v0.4` + setup RTD to generate docs for this branch + all the usual release process etc.)

**Comment by rtavenar** [comment]
> We also have to remember to update the conda recipe with an equivalent apporach.

If I am correct, we only provide conda packages for Python 3 so it should not be an issue.

**Comment by rth** [comment]
> If I am correct, we only provide conda packages for Python 3 so it should not be an issue.

yes, we just have to rember to skip python 2 during build with e.g. https://github.com/conda-forge/scikit-learn-feedstock/blob/3ac704288b5a47d66dbe8b27369b5b917014b230/recipe/meta.yaml#L14

**Comment by rtavenar** [comment]
Now that we have a camera-ready version of the tslearn paper to prepare, I will move item number 7 on the top of the list :)

**Comment by rtavenar** [comment]
I have released v0.4 on PyPI!

I forgot to remove python 2.7 builds before uploading to PyPI. But these should be skipped next time.
